### ISSUE TITLE
Fuzz minizip

### ISSUE THREAD

**Issue by Dor1s** [google/oss-fuzz#1940]
Hi @nmoinvaz,

@akmistry has been doing awesome work to fuzz [minizip in Chromium](https://cs.chromium.org/chromium/src/third_party/minizip/). AFAIK, you might've already received some bug reports. However, to fully leverage our fuzzing capabilities, we would like to invite you to integrate minizip with OSS-Fuzz.

Main advantages of integrating with OSS-Fuzz:

1) We always fuzz tip-of-tree revision here, while in Chromium we use a fixed revision which needs to be manually updated from time to time

2) OSS-Fuzz integration would allow you and other contributors to reproduce crashes in a more convenient way than you might have to do it right now

3) Everything is automated on our end, including bug reporting and fixes verification, which also would make things more convenient for the upstream developers

Also, we offer rewards for open source projects who integrate with OSS-Fuzz, here are some details and links: https://security.googleblog.com/2018/11/a-new-chapter-for-oss-fuzz.html

Please take a look at the [New Project Guide](https://github.com/google/oss-fuzz/blob/master/docs/new_project_guide.md) and let us know if you have any questions!

**Comment by nmoinvaz** [comment]
I have seen @akmistry work's to fuzz minizip in Chromium and it looks great. I have also tried to fix some of the bugs that I've seen come through.

I ran into some problems when trying to setup libFuzzer. I couldn't really find any instructions for Windows. When installing LLVM for Visual Studio, I attempted to use -fsanitize=address option, but it said it was an `unsupported option '-fsanitize=fuzzer' for target 'i386-pc-windows-msvc'`. When looking through the libFuzzer source code, it looks like there is some Windows code in there, but running cmake in the source directory fails with `Unknown CMake command "CHECK_CXX_SOURCE_COMPILES"` even with the latest cmake.  

I am always interested in rewards. What would you want me to do beyond what @akmistry has done? Are you looking for me to move all his fuzzer tests into the minizip repository or something more?

Nathan

**Comment by Dor1s** [comment]
Yes, as @inferno-chromium said over email, "all you need is just move those fuzz targets files to your repo and add < 50 line [build/integration instructions](https://github.com/google/oss-fuzz/blob/master/docs/new_project_guide.md) in OSS-Fuzz github repo ([example](https://github.com/google/oss-fuzz/tree/master/projects/boringssl))". That would be sufficient to get fuzzing coverage on tip-of-tree revision and convenient bug reporting / verification pipeline.

Later on, you can follow more steps from the [Ideal Integration](https://github.com/google/oss-fuzz/blob/master/docs/ideal_integration.md) in order to improve testing of minizip and apply for a higher reward :)

**Comment by nmoinvaz** [comment]
@Dor1s - I spent sometime this weekend figuring out libFuzzer and how to integrate. Here are my changes. Please take a look and let me know if this is correct or if anything is missing. See pull request #1945.

**Comment by Dor1s** [comment]
@nmoinvaz looks like linking doesn't work properly, the binaries are linked against the standalone driver, rather than against `-lFuzzingEngine`: https://github.com/nmoinvaz/minizip/blob/dev/CMakeLists.txt#L671

```cmake
    if (NOT CMAKE_COMPILER_IS_CLANG)
        set(FUZZER_SRC "test/fuzz/standalone.c")
    endif()
```

Do we need to set `CMAKE_COMPILER_IS_CLANG` in https://github.com/google/oss-fuzz/blob/master/projects/minizip/build.sh#L19 ?

**Comment by nmoinvaz** [comment]
@Dor1s - I have updated the cmake clang detection, let me know if that works for you.

**Comment by Dor1s** [comment]
Cool, thanks! Now there is a bunch of link failures:

```
[ 96%] Building C object CMakeFiles/zip_fuzzer.dir/test/fuzz/zip_fuzzer.c.o
[ 98%] Linking C executable unzip_fuzzer
[100%] Linking C executable zip_fuzzer
/usr/lib/gcc/x86_64-linux-gnu/5.4.0/../../../libFuzzingEngine.a(FuzzerDriver.o)/usr/lib/gcc/x86_64-linux-gnu/5.4.0/../../../libFuzzingEngine.a(FuzzerDriver.o): In function `: In function `appendappend':
':
/usr/local/bin/../include/c++/v1/string:/usr/local/bin/../include/c++/v1/string:25172517: undefined reference to `: undefined reference to `std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >::append(char const*, unsigned long)std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >::append(char const*, unsigned long)'
'
/usr/lib/gcc/x86_64-linux-gnu/5.4.0/../../../libFuzzingEngine.a(FuzzerDriver.o)/usr/lib/gcc/x86_64-linux-gnu/5.4.0/../../../libFuzzingEngine.a(FuzzerDriver.o): In function `: In function `__do_call':
__do_call/usr/local/bin/../include/c++/v1/new:':
340/usr/local/bin/../include/c++/v1/new:: undefined reference to `340: undefined reference to `operator delete(void*)operator delete(void*)'
'
```

I think the reason is that we should link using `clang++` even though the project is written in C, because libFuzzer is written in C++.

**Comment by nmoinvaz** [comment]
Is there any way I can reproduce those results? I had tested through docker and also used infra.py script initially during my testing.

**Comment by nmoinvaz** [comment]
I have modified my cmake script to use CXX when linking. Give that a shot.

**Comment by Dor1s** [comment]
```bash
$ python infra/helper.py build_image --no-pull minizip \
    && python infra/helper.py build_fuzzers minizip

...

-- Performing Test check_opt_aes - Success
-- Configuring done
CMake Error: Error required internal CMake variable not set, cmake may be not be built correctly.
Missing variable is:
CMAKE_CXX_LINK_EXECUTABLE
CMake Error: Error required internal CMake variable not set, cmake may be not be built correctly.
Missing variable is:
CMAKE_CXX_LINK_EXECUTABLE
-- Generating done
-- Build files have been written to: /src/minizip
Fuzzers build failed.
```

**Comment by nmoinvaz** [comment]
Thanks for the command. Should be fixed now.

**Comment by nmoinvaz** [comment]
You may require pull request #1948.

**Comment by nmoinvaz** [comment]
I think I've integrated the project with oss-fuzz, and now the relevant bugs should be fixed. Is it possible for you to nominate me for the reward? Thanks.

**Comment by inferno-chromium** [comment]
For rewards, @nmoinvaz - please fill out the form at https://goo.gl/YMLfG2
